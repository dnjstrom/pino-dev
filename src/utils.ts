const repeatString = (amount: number, str: string) =>
  new Array(amount + 1).join(" ");

export const leftpad = (amount: number, str: string, padCharacter = " ") =>
  repeatString(amount, padCharacter) + str;

export const padLinesWithSpaces = (
  amount: number,
  str: string,
  newline: string
) => {
  const spaces = repeatString(amount, " ");
  return str.replace(new RegExp(`${newline}`, "g"), `${newline}${spaces}`);
};

export const getDeep = (keys: string[], obj: any): any => {
  const [key, ...remainingKeys] = keys;

  if (obj == null) {
    return undefined;
  } else if (key == null) {
    return obj;
  }

  return getDeep(remainingKeys, obj[key]);
};

export const setDeep = (obj: any = {}, keys: string[], value: any): any => {
  const [key, ...remainingKeys] = keys;

  if (key == null) {
    return obj;
  } else if (remainingKeys.length === 0) {
    obj[key] = value;
    return obj;
  }

  obj[key] = setDeep(obj[key] ? obj[key] : {}, remainingKeys, value);

  return obj;
};

export const joinWith = (separator: string, items: unknown[]) =>
  items.filter(Boolean).join(separator);

export const words = (...args: unknown[]) => joinWith(" ", args);
