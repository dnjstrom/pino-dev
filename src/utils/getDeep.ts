export const getDeep = <T = unknown>(
  keys: string[],
  obj: unknown,
): T | undefined => {
  const [key, ...remainingKeys] = keys;

  if (obj == null) {
    return undefined;
  } else if (key == null) {
    return obj as T;
  }

  return getDeep(remainingKeys, (obj as Record<string, unknown>)[key]);
};
