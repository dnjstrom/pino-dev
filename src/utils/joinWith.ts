/**
 * Concatenates parameters with separator, omitting any falsy values.
 */

export const joinWith = (
  separator: string,
  items: Array<string | Falsy>,
): string => items.filter(Boolean).join(separator);
