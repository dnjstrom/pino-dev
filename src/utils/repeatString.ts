export const repeatString = (amount: number, str: string): string =>
  new Array(Math.max(amount, 0) + 1).join(str);
