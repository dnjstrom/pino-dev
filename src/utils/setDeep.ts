export const setDeep = <T = unknown>(
  obj: Record<string, unknown>,
  keys: string[],
  value: T
): Record<string, unknown> => {
  const [key, ...remainingKeys] = keys;

  if (key == null) {
    throw new Error(
      `Invalid path "${keys}" when setting "${value}" in \`${JSON.stringify(
        obj
      )}\`.`
    );
  } else if (remainingKeys.length === 0) {
    obj[key] = value;
    return obj;
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const parent = obj as any;
  const child = parent[key] ? parent[key] : {};
  obj[key] = setDeep(child, remainingKeys, value);

  return obj;
};
