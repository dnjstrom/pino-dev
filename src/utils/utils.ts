import { Falsy } from "../types";

const repeatString = (amount: number, str: string): string =>
  new Array(Math.max(amount, 0) + 1).join(str);

export const leftPad = (
  amount: number,
  str: string,
  padCharacter = " "
): string => repeatString(amount, padCharacter) + str;

export const padLinesWithSpaces = (
  amount: number,
  str: string,
  newline: string
): string => {
  const spaces = repeatString(amount, " ");
  return str.replace(new RegExp(`${newline}`, "g"), `${newline}${spaces}`);
};

export const getDeep = <T = unknown>(
  keys: string[],
  obj: unknown
): T | undefined => {
  const [key, ...remainingKeys] = keys;

  if (obj == null) {
    return undefined;
  } else if (key == null) {
    return obj as T;
  }

  return getDeep(remainingKeys, (obj as Record<string, unknown>)[key]);
};

export const setDeep = <T = unknown>(
  obj: Record<string, unknown>,
  keys: string[],
  value: T
): Record<string, unknown> => {
  const [key, ...remainingKeys] = keys;

  if (key == null) {
    throw new Error(
      `Invalid path "${keys}" when setting "${value}" in \`${JSON.stringify(
        obj
      )}\`.`
    );
  } else if (remainingKeys.length === 0) {
    obj[key] = value;
    return obj;
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const parent = obj as any;
  const child = parent[key] ? parent[key] : {};
  obj[key] = setDeep(child, remainingKeys, value);

  return obj;
};

/**
 * Concatenates parameters with separator, omitting any falsy values.
 */
export const joinWith = (
  separator: string,
  items: Array<string | Falsy>
): string => items.filter(Boolean).join(separator);
