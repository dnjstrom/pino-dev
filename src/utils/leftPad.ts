import { repeatString } from "./repeatString";

export const leftPad = (
  amount: number,
  str: string,
  padCharacter = " "
): string => repeatString(amount, padCharacter) + str;
