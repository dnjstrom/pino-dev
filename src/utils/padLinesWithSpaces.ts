import { repeatString } from "./repeatString";

export const padLinesWithSpaces = (
  amount: number,
  str: string,
  newline: string
): string => {
  const spaces = repeatString(amount, " ");
  return str.replace(new RegExp(`${newline}`, "g"), `${newline}${spaces}`);
};
