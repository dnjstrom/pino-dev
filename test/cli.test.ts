import { execSync } from "child_process";
import { examples } from "./utils/examples";
import path from "path";
import packageJson from "../package.json";

const BIN_PATH = path.resolve(__dirname, "..", packageJson.bin["pino-dev"]);

const runCli = (input: string, flags = "") =>
  execSync(`FORCE_COLOR=0 ${BIN_PATH} ${flags}`, {
    input,
  }).toString();

describe("cli", () => {
  it("handles all examples correctly", () => {
    const input = examples.join("\n");
    expect(runCli(input)).toMatchSnapshot();
  });

  it("applies command line flags", () => {
    const input = `{"foo": "hello", "time": 1612899089589}`;
    const flags = `--time-format 'HH:mm:ss' --newline '\r\n' --property-map '{ "msg": "foo" }'`;
    const output = runCli(input, flags);
    expect(output).toMatchInlineSnapshot(`
      "19:31:29: hello
      "
    `);
  });
});
