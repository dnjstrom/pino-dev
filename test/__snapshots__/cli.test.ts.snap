// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli handles all examples correctly 1`] = `
"miscellaneous text
02:56:15 Trace: An info message
02:56:15 Debug: A debug message
02:56:15 Info: An info message
02:56:15 Warn: A warning message
02:56:15 Error: An error message
02:56:15 Fatal: A fatal message
02:56:15 Info: 	    A message with leading whitespace
02:56:15 Info: A message with ending whitespace    	
02:56:15 Info: A multi-line message. Lorem ipsum
               dolor sit amet, consectetur adipiscing elit. Vivamus vitae orci volutpat,
               finibus augue vitae, pellentesque nibh. Maecenas lacus erat, maximus sit amet ligula sed,
               porttitor tincidunt felis. Donec dapibus eget est eu fermentum. Sed eu augue turpis.
               Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia curae;
               Vivamus vehicula ante nisi, id porta odio blandit ut.
02:56:15 Info: A message with an object parameter
02:56:15 Error: An error happened somehow
    Error: Oh noes!
    at /some/path/to/file:23:98
    at step (/some/path/to/file:23:98)
    at Object.next (/some/path/to/file:23:98)
    at /some/path/to/file:23:98
    at new Promise (<anonymous>)
    at __awaiter (/some/path/to/file:23:98)
    at generateExamples (/some/path/to/file:23:98)
    at Object.<anonymous> (/some/path/to/file:23:98)
    at Module._compile (internal/some/path/to/file:23:98)
    at Module.m._compile (/some/path/to/file:23:98)
02:56:15 Info: Meta data via child logger
02:56:15 [My Logger] Info: A logger with a name set
02:56:15 Info: A logger without the base child logger
02:56:15 Info: something else
02:56:15 Info: GET /greet/world 200 123ms
02:56:15 [My Logger > my-test:namespace] Trace: A debug message
"
`;
